### ISSUE TITLE
support for kernel 4.7

### ISSUE THREAD

**Issue by Finkregh** [archzfs/archzfs#63]
Hi,

archlinux switched to a 4.7 kernel some days ago :-)

**Comment by storrgie** [comment]
Coming in to +1 this as well

**Comment by stevleibelt** [comment]
Hope everything is fine (like enjoying a deserved holiday :-)).

After dealing with important stuff, it would be great if you would enable release the support for linux kernel 4.7.x.

**Comment by demizer** [comment]
Hi everyone. We'll have to wait til a stable release is made upstream. Not sure how long that will be. FWIW, this is the first time this has happened in the last year and a half. Seems like they are putting the final touches on a stable release. The last stable release was in May 2016.

Until then it is possible to use either archzfs-linux-git or archzfs-linux-lts.

**Comment by stevleibelt** [comment]
@demizer 

thank you very much for clearing this up.
I don't wanted to put either pressure nor bad feelings on your shoulders. As you pointed out, this has never happened bevor and there is an alternative available.
While you are here, thanks for your work :-).

**Comment by Finkregh** [comment]
hmm, perhaps we should highlight the usage of **-lts** in the README :-)

**Comment by crocket** [comment]
@demizer What do you mean by a stable release? A stable release of zfs or linux?

**Comment by demizer** [comment]
@Finkregh will do
@crocket A stable release of ZFSonLinux at zfsonlinux.org that supports linux 4.7

**Comment by stevleibelt** [comment]
@demizer 

To validate that I am doing something wrong. Does archzfs-linux-git support the arch linux kernel 4.7.1-1?
If I try to create an archiso, I get the following errors:

```
[...]
[mkarchiso] INFO: Installing packages to 'work/x86_64/airootfs/'...
==> Creating install root at work/x86_64/airootfs
==> Installing packages to work/x86_64/airootfs
:: Synchronizing package databases...
 core is up to date
 extra is up to date
 community is up to date
 archzfs is up to date
:: There are 4 members in group archzfs-linux-git:
:: Repository archzfs
   1) spl-linux-git  2) spl-utils-linux-git  3) zfs-linux-git  4) zfs-utils-linux-git

Enter a selection (default=all): 
resolving dependencies...
warning: cannot resolve "linux=4.7", a dependency of "spl-linux-git"
warning: cannot resolve "linux=4.7", a dependency of "spl-linux-git"
warning: cannot resolve "spl-linux-git", a dependency of "zfs-linux-git"
warning: cannot resolve "linux=4.7", a dependency of "spl-linux-git"
warning: cannot resolve "spl-linux-git", a dependency of "zfs-utils-linux-git"
warning: cannot resolve "linux=4.7", a dependency of "zfs-utils-linux-git"
warning: cannot resolve "zfs-utils-linux-git", a dependency of "zfs-linux-git"
warning: cannot resolve "linux=4.7", a dependency of "zfs-linux-git"
warning: cannot resolve "linux=4.7", a dependency of "spl-linux-git"
warning: cannot resolve "spl-linux-git", a dependency of "zfs-utils-linux-git"
warning: cannot resolve "linux=4.7", a dependency of "zfs-utils-linux-git"
:: The following packages cannot be upgraded due to unresolvable dependencies:
      spl-linux-git  zfs-linux-git  zfs-utils-linux-git

:: Do you want to skip the above packages for this upgrade? [y/N] 
error: failed to prepare transaction (could not satisfy dependencies)
:: spl-linux-git: requires linux=4.7
:: spl-linux-git: requires linux=4.7
:: zfs-linux-git: requires spl-linux-git
:: spl-linux-git: requires linux=4.7
:: zfs-utils-linux-git: requires spl-linux-git
:: zfs-utils-linux-git: requires linux=4.7
:: zfs-linux-git: requires zfs-utils-linux-git
:: zfs-linux-git: requires linux=4.7
:: spl-linux-git: requires linux=4.7
:: zfs-utils-linux-git: requires spl-linux-git
:: zfs-utils-linux-git: requires linux=4.7
==> ERROR: Failed to install packages to new root
[...]
```

**Comment by johnramsden** [comment]
So are we still waiting on a ZOL release for zfs-linux to be updated?

**Comment by demizer** [comment]
@johnramsden yes

**Comment by johnramsden** [comment]
Do we have any way of finding out how long it will be? 

In unrelated, I attempted to switch to zfs-linux-git, and ended up getting errors related to linux not being able to upgrade due to dependencies. What should I do about this?

```
resolving dependencies...
warning: ignoring package linux-4.7.2-1
warning: cannot resolve "linux=4.7.2", a dependency of "spl-linux-git"
warning: cannot resolve "spl-linux-git", a dependency of "zfs-linux-git"
warning: ignoring package linux-4.7.2-1
warning: cannot resolve "linux=4.7.2", a dependency of "spl-linux-git"
warning: cannot resolve "spl-linux-git", a dependency of "zfs-utils-linux-git"
warning: ignoring package linux-4.7.2-1
warning: cannot resolve "linux=4.7.2", a dependency of "zfs-utils-linux-git"
warning: cannot resolve "zfs-utils-linux-git", a dependency of "zfs-linux-git"
warning: ignoring package linux-4.7.2-1
warning: cannot resolve "linux=4.7.2", a dependency of "zfs-linux-git"
:: The following package cannot be upgraded due to unresolvable dependencies:
      zfs-linux-git
```

**Comment by lnicola** [comment]
@johnramsden I guess you can follow zfsonlinux/zfs#5025.

I had the same issue updating to the `-git` version and I ended up using `-d` or `-dd` to tell `pacman` to ignore dependencies. Basically, I removed `zfs-linux`, `spl-linux`, `zfs-utils-linux` and `spl-utils-linux` and then installed the corresponding `-git` versions. Just hope the power doesn't go off 😄 . This is because I don't think `pacman` can remove a package and install another one in the same transaction.

If you update, make sure to enable the new `systemd` units. I knew people had trouble, but I didn't do it and I ended up with only `/` mounted and an unusable system. One way that might work is to run `systemctl daemon-reload` and `systemctl enable zfs-import-cache zfs-import-scan zfs-mount zfs-share zfs-zed`, though I'm not sure it's the best and how those relate to `zfs.target`. Actually, that's because I don't know much about `systemd` targets.

**Comment by jacksonatl** [comment]
I have a cookie for anyone with experience who would do a quite write up on best practices when changing kernels, particularly going from current to LTS. I'm sure it's pretty straight forward, but one question I can't find a clear answer on is, should you export the file system before changing kernels or is it just a matter of replacing modules after reboot?

**Comment by demizer** [comment]
@johnramsden I have not heard anything about when support for 4.7 will be released. All I know is that they are currently finalizing a big change that they want to make sure is fully tested before making a new stable release. Also, regarding the dependency issue, could you make a new ticket out of that and we'll see about getting it fixed. Thanks.

In other news, a side affect of no upstream support for 4.7 is that the archiso is currently using 4.7, so it is not possible to boot into the archiso to install ZFS. I have put together a custom archiso for booting into linux-lts that can be seen in the "initscript_refactor" branch in the ```testing/tests/archzfs-qemu-lts-test-01-root-bootfs/archiso``` directory. For those that may want to use it to install archzfs, see the README in the same directory.

@jacksonatl Great idea! Also, I plan on investigating if it is possible to install archzfs-linux and archzfs-linux-lts on the same host without file conflicts once I get a test environment stood up. We're almost there! Thanks.

**Comment by lnicola** [comment]
> should you export the file system before changing kernels
I see no reason to do that. You'll be running the same version of ZFS after the kernel change.

**Comment by nedbass** [comment]
ZFSonLinux version 0.6.5.8 with support for 4.7 kernels is now available at https://github.com/zfsonlinux/zfs/releases/tag/zfs-0.6.5.8

Thanks for your patience while we finalized this release.

**Comment by stevleibelt** [comment]
Thank you for your work! :+1: :+1: :+1: :+1: :+1: :+1:

**Comment by lej1337** [comment]
bullshit, I still can't get past upgrading to the 4.7 kernels, which
requires 4.6.

**Comment by lnicola** [comment]
@lej1337 See my comment above: https://github.com/archzfs/archzfs/issues/63#issuecomment-244588604

**Comment by lej1337** [comment]
I shouldn't have to do all of that. The main package isn't even updated to
get past the 4.6 dependencies, which they claim it is. Please update so
folks like me who aren't so smart in this stuff can get a working live cd
with zfs.

On Sat, Sep 10, 2016 at 11:35 AM, Laurentiu Nicola <notifications@github.com

**Comment by haraldkoch** [comment]
@lej1337 - that comment was from a ZFS developer saying their stable release is available. It was not a comment from the Arch ZFS maintainer, aka @demizer.

**Comment by johnramsden** [comment]
@nedbass Thanks!

So we should expect an update to zfs-linux in the next few days? Does that mean 4.7 support for zfs-linux is ready?

**Comment by storrgie** [comment]
> @lej1337 - that comment was from a ZFS developer saying their stable release is available. It was not a comment from the Arch ZFS maintainer, aka @demizer.

> folks like me who aren't so smart in this stuff can get a working live cd
with zfs.

@lej1337 get out of here

**Comment by demizer** [comment]
He everyone. I have started the build / test for 0.6.5.8 from https://github.com/zfsonlinux/zfs/releases/latest

Should be in a day or two hopefully. This version has changes to the systemd scripts that I need to verify, so it will be a little while more before I can make a release.

**Comment by stevleibelt** [comment]
@demizer 

Awesome! Thanks for your work.
@lej1337 move along.

**Comment by aut0** [comment]
@lej1337 your demeanor is totally uncalled for and really ungrateful. You should reconsider how you treat people who put in a lot of their time so we can have nice things. They don't even get payed for it. "Being not so smart" is not an excuse for behaving like this here.

**Comment by demizer** [comment]
I almost forgot to thank @nedbass for letting us know a stable release has been made. Thanks!

**Comment by georgyo** [comment]
I have built the packages using your tools, and have installed 0.6.5.8 and rebooted after enabling the new ZFS services. I can report that everything is working well.

Thanks for creating awesome build scripts!

**Comment by demizer** [comment]
@georgyo Awesome! Thanks!

**Comment by storrgie** [comment]
I don't mean to be silly and re-open, but with the closure of this ticket is the expected behavior to be able to `pacman -Syu` and pull both the new kernel and new zfs packages? As of now I'm still seeing conflicts:

```
resolving dependencies...
warning: cannot resolve "linux=4.7.5", a dependency of "spl-linux"
warning: cannot resolve "linux=4.7.5", a dependency of "spl-linux"
warning: cannot resolve "linux=4.7.5", a dependency of "spl-linux"
warning: cannot resolve "linux=4.7.5", a dependency of "zfs-utils-linux"
warning: cannot resolve "linux=4.7.5", a dependency of "zfs-linux"
warning: cannot resolve "linux=4.7.5", a dependency of "spl-linux"
warning: cannot resolve "linux=4.7.5", a dependency of "zfs-utils-linux"
:: The following packages cannot be upgraded due to unresolvable dependencies:
      spl-linux  zfs-linux  zfs-utils-linux
```

I am on 4.6.4-1-ARCH

**Comment by johnramsden** [comment]
@storrgie You might want to try ```pacman -S zfs-linux```

**Comment by demizer** [comment]
@storrgie You got this error because at the time of your update attempt from kernel 4.6.4-1-ARCH to 4.7.6-1-ARCH, zfs-linux was not yet updated for 4.7.6. The zfs package in the archzfs repo at the time had a hard dependency on 4.7.5 (which is greater than your current kernel version), but the linux=4.7.5 package was no longer in the official arch repos.

I have updated the zfs-linux packages and you should be able to update without errors.
